testing/src/scenario/context.py

```diff
@@ -355,7 +355,7 @@ def action(
         return _Event(f"{name}_action", action=_Action(name, **kwargs))
 
 
-class Context:
+class Context(Generic[CharmType]):
```

Comment:
I wonder if we also ought to be pushing this through to the `self.ops` object, so `ops_main_mock.Ops` has it as well. I think that means that the cast on line 94 could be an "assert not None" line instead - I'm not sure if there are other benefits as well. @james-garner-canonical any thoughts on that?

Reply:
Doing that would probably be correct as I see that the `Ops` class has a `charm_spec` argument that's parameterised on `charm_spec` (`_CharmSpec[CharmType]`). Luckily this will already be correctly parameterised when constructed since `Ops` is initialised via the (now generic) `Context`. So it should be as simple as changing `class Ops` to `class Ops(Generic[CharmType])` and changing the type annotation in `Manager.__init__`.

Line 90 should probably be `if self.ops is None` rather than `if not self.ops`, right? Then you should be able to eliminate the `cast` without needing an `assert`.

Reply:
Is this what you had in mind? a4a37d8b